### PR TITLE
Referral page

### PR DESCRIPTION
**Pull request by cpariwat**

### CODE REVIEW

File: app/controllers/users/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class Users::RegistrationsController < Devise::RegistrationsController
+  before_action :configure_sign_up_params, only: [:create]
```
Comment (citizen428) [inline app/controllers/users/registrations_controller.rb]: If you have a `before_action` that you only need in one method, I would prefer if you explicitly call it there. This makes it much easier to fully understand the action without hopping through the controller.

File: app/models/concerns/invitable.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+module Invitable
+  extend ActiveSupport::Concern
```
Comment (citizen428) [inline app/models/concerns/invitable.rb]: I'm not a big fan of concerns. All they do is hide how big a model really is. It's the equivalent of taking stuff from a messy desk and throwing it in a drawer. The mess is still there, you just don't see it :wink: 

I'd prefer if invitations could be a separate model, where you record both the inviter and the invitee. Ideally user would not need to change when adding or removing invitations. Also you can have a new token for every invite, it doesn't matter, so you can record it with the invitation itself (i.e. create the token when creating the invite). So the attributes you'd need would be `inviter_id`, `invitee_id`, `token` and `accepted` (boolean).

File: config/routes.rb [hunk 3]
```diff
@@ -1,7 +1,14 @@
...
 
-  unauthenticated :user do
+  devise_scope :user do
+    authenticate :user do
+      get '/pages/referral' => 'pages#show', id: 'referral', format: false
+    end
+  end
+
+  unauthenticated :users do
+    get '/referral/:invite_token' => 'invites#show', as: :referral, format: false
```
Comment (citizen428) [inline config/routes.rb]: Prefer using the routing DSL (also above):

```ruby
    resources :referral, only: [:show], param: :invite_token, format: false
```

File: app/controllers/invites_controller.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class InvitesController < ApplicationController
+  skip_before_action :authenticate_user!
+
+  def show
+    if user_signed_in?
+      redirect_to root_path
```
Comment (citizen428) [inline app/controllers/invites_controller.rb]: `authenticated_root_path`

File: app/models/subscription.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  def friends_joined
+    invitees.count
+  end
+
+  def rank
+    Subscription.most_friends_joined.pluck(:id).index(id) + 1
+  end
+
+  def days_left
+    return 1000 if token_expire_at.nil?
```
Comment (citizen428) [inline app/models/subscription.rb]: Please don't just use random numbers like this. If you have to at least assign them to a constant like `MAX_DAYS` or something and use that here.

File: app/models/subscription.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  def self.token_valid?(token)
+    subscription = from_token(token).first
+    subscription.days_left.positive? && subscription.token_valid
+  end
+
+  def friends_joined
+    invitees.count
+  end
+
+  def rank
```
[the remark this answers is not in the document]
Comment (cpariwat) [reply app/models/subscription.rb]: Thank you 👍   I wasn't taking in account for more generic leaderboard scenario.

File: app/models/subscription.rb (hunk 5 above)
Comment (cpariwat) [reply app/models/subscription.rb]: I'll make a temporary constant and set it to 60 days until we have more information on this.

**Pull request merged**